File: 算法与数据结构/golang实现数据结构/DataStructAndAlgorithm/queue/queue/queue.go
package queue

import (
	"DataStructAndAlgorithm/queue/array"
	"fmt"
)

//在已实现的动态数组基础上通过接口实现队列
type ArrayQueue struct {
	array *array.Array
}

// 新建ArrayQueue对象实例，指定容量
func NewArrayQueue(capacity int) *ArrayQueue {
	arr := array.NewArray(capacity)
	queue := ArrayQueue{array: arr}
	return &queue
}

// 新建ArrayQueue对象实例，使用默认容量
func NewArrayQueue1() *ArrayQueue {
	arr := array.NewArray1()
	queue := ArrayQueue{array: arr}
	return &queue
}

// 元素入队
func (q *ArrayQueue) Enqueue(e interface{}) error {
	return q.array.AddLast(e)
}

// 元素出队
func (q *ArrayQueue) Dequeue() (interface{}, error) {
	return q.array.RemoveFirst()
}

// 查询队首元素
func (q *ArrayQueue) GetFront() (interface{}, error) {
	return q.array.GetFirst()
}

// 获取队列中元素数量
func (q *ArrayQueue) GetSize() int {
	return q.array.GetSize()
}

// 获取队列容量
func (q *ArrayQueue) GetCapacity() int {
	return q.array.GetCapacity()
}

// 判断队列是否为空
func (q *ArrayQueue) IsEmpty() bool {
	return q.array.IsEmpty()
}

// 格式化输出队列元素信息
func (q *ArrayQueue) String() string {
	res := fmt.Sprintf("Queue: front [")
	for i:=0; i<q.GetSize(); i++ {
		data, _ := q.array.Get(i)
		res += fmt.Sprintf("%v", data)
		if i != q.GetSize()-1 {
			res += ", "
		}
	}
	res += "] tail"
	return res
}

File: 算法与数据结构/python实现数据结构/queue_/queue.py
# coding:utf-8
from array_.array import Array


class ArrayQueue(object):
    def __init__(self, capacity):
        self._array = Array(capacity)

    def enqueue(self, e):
        """
        元素入队
        :param e:
        :return:
        """
        return self._array.add_last(e)

    def dequeue(self):
        """
        元素出队
        :return:
        """
        return self._array.remove_first()

    def get_front(self):
        """
        查询队顶元素
        :return:
        """
        return self._array.get_first()

    def get_size(self):
        """
        获取队列中元素个数
        :return:
        """
        return self._array.get_size()

    def get_capacity(self):
        """
        获取队列的容量
        :return:
        """
        return self._array.get_capacity()

    def is_empty(self):
        """
        判断队列是否为空
        :return:
        """
        return self._array.is_empty()

    def __repr__(self):
        res = "Queue: front ["
        for i in range(self._array.get_size()):
            res += "{0}".format(self._array.get(i))
            if i != self._array.get_size() - 1:
                res += ", "
        res += "] tail"

        return res


def test_array_queue():
    q = ArrayQueue(10)
    for i in range(10):
        q.enqueue(i)
    print(q)
    q.dequeue()
    print(q)


if __name__ == "__main__":
    test_array_queue()
